scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/data/TaskProActiveDataspaces.java

```diff
@@ -285,162 +318,241 @@ protected void logDataspacesStatus(String message, DataspacesStatusLevel level)
+
+            DataSpacesFileObject target = destination.resolveFile(relativePath);
+
+            if (target.isFolder()) {
+                target.createFolder();
+            } else if (target.isFile()) {
+                target.getParent().createFolder();
             }
 
-            StringBuilder exceptionMsg = new StringBuilder();
-            String nl = System.lineSeparator();
-            for (Future<Boolean> f : transferFutures) {
-                try {
-                    f.get();
-                } catch (InterruptedException | ExecutionException e) {
-                    logger.error("", e);
-                    exceptionMsg.append(StackTraceUtil.getStackTrace(e)).append(nl);
-                }
+            DataSpacesFileObject oldFileObject = filesToCopy.put(relativePath, fileObject);
+            if (oldFileObject != null) {
+                String message = fileObject.getRealURI()
+                        + " will be copied instead of " + oldFileObject.getRealURI() + ".\n "
+                        + "Precedence order is output space, input space, user space, global space.";
+                logger.warn(message);
+                logDataspacesStatus(message, DataspacesStatusLevel.WARNING);
             }
-            if (exceptionMsg.length() > 0) {
-                toBeThrown = new FileSystemException(
-                        "Exception(s) occurred when transferring input files : " + nl + exceptionMsg.toString());
+        }
+    }
+
+    private void handleResults(List<Future<Boolean>> transferFutures) throws FileSystemException {
+
+        StringBuilder message = new StringBuilder();
+        String nl = System.lineSeparator();
+
+        for (Future<Boolean> future : transferFutures) {
+            try {
+                future.get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Exception while fetching dataspace transfer result ", e);
+                message.append(StackTraceUtil.getStackTrace(e)).append(nl);
             }
+        }
 
-            if (toBeThrown != null) {
-                throw toBeThrown;
+        if (message.length() > 0) {
+            throw new FileSystemException(
+                    "Exception(s) occurred when transferring input file: " + nl + message.toString());
+        }
+    }
+
+    private String relativize(String inputSpaceUri, DataSpacesFileObject fileObject) {
+        return fileObject.getVirtualURI().replaceFirst(inputSpaceUri + "/?", "");
+    }
+
+    private String virtualResolve(DataSpacesFileObject dataSpacesFileObject) {
+        if (dataSpacesFileObject == null) {
+            return "";
+        } else {
+            return dataSpacesFileObject.getVirtualURI();
+        }
+    }
+
+    private FileSystemException findFilesToCopyFromInputToScratch(List<InputSelector> inputSelectors,
+            ArrayList<DataSpacesFileObject> inResults, ArrayList<DataSpacesFileObject> outResults,
```

Comment:
For sure we can but what does it bring to this *private* method? It is not exposed in a public API and the list implementation is not going to change.